### PR TITLE
[Impeller] Clamp texture decompression to maximum supported texture size.

### PR DESCRIPTION
**Pull request by magicianA**
fixes https://github.com/flutter/flutter/issues/108929
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/renderer/allocator.h [hunk 1]
```diff
@@ -69,6 +69,8 @@ class Allocator {
   std::shared_ptr<DeviceBuffer> CreateBufferWithCopy(
       const fml::Mapping& mapping);
 
+  virtual ISize GetMaxTextureSizeSupported() = 0;
```
Comment (chinmaygarde) [inline impeller/renderer/allocator.h:72]: Add a `const` here?

`virtual ISize GetMaxTextureSizeSupported() const = 0;`

File: impeller/renderer/backend/vulkan/allocator_vk.cc [hunk 2]
```diff
@@ -143,5 +143,9 @@ std::shared_ptr<DeviceBuffer> AllocatorVK::CreateBuffer(StorageMode mode,
   return std::make_shared<DeviceBufferVK>(length, mode, context_,
                                           std::move(device_allocation));
 }
-
+ISize AllocatorVK::GetMaxTextureSizeSupported() {
+  // source :
+  // https://registry.khronos.org/vulkan/specs/1.2-extensions/html/vkspec.html#limits-minmax
+  return {4096, 4096};
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/allocator_vk.cc:149]: I believe these are just required limits. The actual supported may be much larger. Let's add a TODO with a bug to figure this out for Vulkan later.

File: impeller/renderer/backend/vulkan/allocator_vk.h [hunk 3]
```diff
@@ -45,6 +45,8 @@ class AllocatorVK final : public Allocator {
       StorageMode mode,
       const TextureDescriptor& desc) override;
 
+  ISize GetMaxTextureSizeSupported() override;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/allocator_vk.h:48]: Here and elsewhere, the overridden functions need to have a comment that mentions the base class that contains the method being overriden. So `// |Allocator|` here like for the functions above.

File: impeller/renderer/backend/metal/allocator_mtl.mm [hunk 4]
```diff
@@ -173,4 +173,31 @@ static MTLStorageMode ToMTLStorageMode(StorageMode mode,
...
 
+ISize AllocatorMTL::GetMaxTextureSizeSupported() {
+  // Since Apple didn't expose API for us to get the max texture size, we have
+  // to use hardcoded data from
+  // https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf
+  // According to the feature set table, there are two supported max sizes :
+  // 16384 and 8192 for devices flutter support. The former is used on macs and
+  // latest ios devices. The latter is used on old ios devices.
+  if (@available(macOS 10.15, iOS 13, tvOS 13, *)) {
+    if ([device_ supportsFamily:MTLGPUFamilyApple3] ||
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/allocator_mtl.mm:184]: Nit: Missing braces around `if` statements.

File: impeller/renderer/backend/metal/allocator_mtl.mm [hunk 5]
```diff
@@ -173,4 +173,31 @@ static MTLStorageMode ToMTLStorageMode(StorageMode mode,
...
 }
 
+ISize AllocatorMTL::GetMaxTextureSizeSupported() {
+  // Since Apple didn't expose API for us to get the max texture size, we have
+  // to use hardcoded data from
+  // https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf
+  // According to the feature set table, there are two supported max sizes :
+  // 16384 and 8192 for devices flutter support. The former is used on macs and
+  // latest ios devices. The latter is used on old ios devices.
+  if (@available(macOS 10.15, iOS 13, tvOS 13, *)) {
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/allocator_mtl.mm:183]: Really not sure how expensive the determination of these limits is, but perhaps figure these out in the ctor of the allocator? See the [UMA and memoryless targets support check here](https://github.com/flutter/engine/blob/2c3efa0beb248c6afd4a0a5eb66ea3fb67a1f9ed/impeller/renderer/backend/metal/allocator_mtl.mm#L59).

File: lib/ui/painting/image_decoder_impeller.h [hunk 6]
```diff
@@ -33,7 +33,9 @@ class ImageDecoderImpeller final : public ImageDecoder {
 
   static std::shared_ptr<SkBitmap> DecompressTexture(
       ImageDescriptor* descriptor,
-      SkISize target_size);
+      SkISize target_size,
+      int32_t max_width,
```
Comment (chinmaygarde) [inline lib/ui/painting/image_decoder_impeller.h:37]: Why not just ISize instead of the two args?

**Review by chinmaygarde — COMMENTED** [review]
Mostly LGTM with some comments and nits. Thanks a ton for this!

File: impeller/renderer/allocator.h (hunk 1 above)
Comment (magicianA) [reply impeller/renderer/allocator.h]: Done

File: impeller/renderer/backend/vulkan/allocator_vk.cc (hunk 2 above)
Comment (magicianA) [reply impeller/renderer/backend/vulkan/allocator_vk.cc:152]: Done.

File: impeller/renderer/backend/vulkan/allocator_vk.h (hunk 3 above)
Comment (magicianA) [reply impeller/renderer/backend/vulkan/allocator_vk.h]: Done.

File: impeller/renderer/backend/metal/allocator_mtl.mm (hunk 4 above)
Comment (magicianA) [reply impeller/renderer/backend/metal/allocator_mtl.mm]: Done.

File: impeller/renderer/backend/metal/allocator_mtl.mm (hunk 5 above)
Comment (magicianA) [reply impeller/renderer/backend/metal/allocator_mtl.mm]: Done.

File: lib/ui/painting/image_decoder_impeller.h (hunk 6 above)
Comment (magicianA) [reply lib/ui/painting/image_decoder_impeller.h]: Done.

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**